File: rocketenv/sample_return/physics.py
"""Pure payload-aware rigid-body dynamics for the sample-return vehicle."""

from __future__ import annotations

import math

import numpy as np

from ..config import Config
from ..physics import FUEL, OMEGA, THETA, VX, VY, X, Y
from .vehicle import VehicleModel


def thrust_direction(theta: float, phi: float) -> np.ndarray:
    """World-frame thrust unit vector for attitude ``theta`` and gimbal ``phi``."""
    return np.array(
        [-math.sin(theta + phi), math.cos(theta + phi)], dtype=np.float64
    )


def thrust_torque(vehicle: VehicleModel, thrust: float, phi: float) -> float:
    """Torque about the combined COM from thrust applied at the engine."""
    engine_x, engine_y = vehicle.engine_offset_from_com_body
    force_x = -thrust * math.sin(phi)
    force_y = thrust * math.cos(phi)
    return float(engine_x * force_y - engine_y * force_x)


def step_dynamics(
    state: np.ndarray,
    action: np.ndarray,
    cfg: Config,
    vehicle: VehicleModel,
) -> np.ndarray:
    """Advance one fixed timestep without mutating any input.

    Engine thrust is rated from dry mass, while gravity and translational
    acceleration use total mass.  Fuel remains massless, matching the original
    simulator.
    """
    x, y, vx, vy, theta, omega, fuel = state

    throttle = min(max(float(action[0]), 0.0), 1.0)
    gimbal_cmd = min(max(float(action[1]), -1.0), 1.0)
    phi = gimbal_cmd * cfg.phi_max

    dry_max_thrust = cfg.twr * vehicle.dry_mass * cfg.g
    thrust = (
        throttle * dry_max_thrust * cfg.thrust_multiplier if fuel > 0.0 else 0.0
    )

    total_mass = vehicle.total_mass
    direction = thrust_direction(theta, phi)
    fx = thrust * direction[0] + cfg.wind_x + cfg.wind_gust_x
    fy = thrust * direction[1] - total_mass * cfg.g
    if cfg.drag_coeff > 0.0:
        speed = math.hypot(vx, vy)
        fx -= cfg.drag_coeff * speed * vx
        fy -= cfg.drag_coeff * speed * vy

    tau = thrust_torque(vehicle, thrust, phi)

    # Semi-implicit Euler, kept in the same order as the legacy dynamics.
    vx += (fx / total_mass) * cfg.dt
    vy += (fy / total_mass) * cfg.dt
    x += vx * cfg.dt
    y += vy * cfg.dt
    omega += (tau / vehicle.total_inertia) * cfg.dt
    theta += omega * cfg.dt

    fuel = max(0.0, fuel - cfg.burn_rate * throttle * cfg.dt)

    return np.array([x, y, vx, vy, theta, omega, fuel], dtype=np.float64)


__all__ = ["step_dynamics", "thrust_direction", "thrust_torque"]

File: web/src/sim/physics.ts
/* Port of rocketenv/sample_return/physics.py.
 *
 * Line-for-line, including the order of the semi-implicit Euler updates —
 * velocity before position, and omega before theta. JavaScript numbers are
 * IEEE-754 float64 and the Python side is float64 throughout, so this is
 * expected to agree to near machine precision. test/parity.test.ts checks it
 * against the recorded trajectory.
 */

import type { StateVec, Vec2 } from "../types";
import { SIM } from "./config";
import type { VehicleModel } from "./vehicle";

export function thrustDirection(theta: number, phi: number): Vec2 {
  return [-Math.sin(theta + phi), Math.cos(theta + phi)];
}

/** Torque about the combined COM from thrust applied at the engine. */
export function thrustTorque(
  vehicle: VehicleModel,
  thrust: number,
  phi: number,
): number {
  const [ex, ey] = vehicle.engineOffsetFromCom;
  const fx = -thrust * Math.sin(phi);
  const fy = thrust * Math.cos(phi);
  return ex * fy - ey * fx;
}

const clamp = (v: number, lo: number, hi: number) =>
  Math.min(Math.max(v, lo), hi);

/** Advance one fixed timestep. Pure: returns a new array. */
export function stepDynamics(
  state: StateVec,
  action: readonly [number, number],
  vehicle: VehicleModel,
): StateVec {
  let [x, y, vx, vy, theta, omega, fuel] = state;

  const throttle = clamp(action[0], 0, 1);
  const phi = clamp(action[1], -1, 1) * SIM.phiMax;

  const dryMaxThrust = SIM.twr * vehicle.dryMass * SIM.g;
  const thrust = fuel > 0 ? throttle * dryMaxThrust * SIM.thrustMultiplier : 0;

  const totalMass = vehicle.totalMass;
  const dir = thrustDirection(theta, phi);
  let fx = thrust * dir[0] + SIM.windX;
  let fy = thrust * dir[1] - totalMass * SIM.g;
  if (SIM.dragCoeff > 0) {
    const speed = Math.hypot(vx, vy);
    fx -= SIM.dragCoeff * speed * vx;
    fy -= SIM.dragCoeff * speed * vy;
  }

  const tau = thrustTorque(vehicle, thrust, phi);

  vx += (fx / totalMass) * SIM.dt;
  vy += (fy / totalMass) * SIM.dt;
  x += vx * SIM.dt;
  y += vy * SIM.dt;
  omega += (tau / vehicle.totalInertia) * SIM.dt;
  theta += omega * SIM.dt;

  fuel = Math.max(0, fuel - SIM.burnRate * throttle * SIM.dt);

  return [x, y, vx, vy, theta, omega, fuel];
}
